projects/imagenet/experiments/sparse_r1.py

```diff
@@ -0,0 +1,196 @@
+
+The sparse 100 configuration here gets between 80.5 to 82% in top-1 accuracy, after 60
+epochs and has about 77.5% average weight sparsity.
+
+The sparse 1000 configuration, with the same weight sparsities, gets about 72.23% in
+top-1 accuracy after 120 epochs.
+
+"""
+
+
+def my_auto_sparse_conv_params(in_channels, out_channels, kernel_size):
+    """
+    Custom weight params.
+    :return: a dict to pass to `SparseWeights2d`
+    """
+    weights_per_channel = kernel_size * kernel_size * in_channels
+    if weights_per_channel < 100:
+        weights_density = 0.7
+
+    elif weights_per_channel < 200:
+        weights_density = 0.5
+
+    elif weights_per_channel < 500:
+        weights_density = 0.4
+
+    elif weights_per_channel < 1000:
+        weights_density = 0.3
+
+    elif weights_per_channel < 2000:
+        weights_density = 0.2
+
+    elif weights_per_channel < 4000:
+        weights_density = 0.2
+
+    else:
+        weights_density = 0.15
+
+    return dict(
+        weight_sparsity=weights_density,
+    )
+
+
+def my_auto_sparse_activation_params(in_channels, out_channels, kernel_size):
+    """
+    A custom auto sparse params function.
+    :return: a dict to pass to `KWinners2d` as params.
+    """
+
+    percent_on = 1.0
+    if kernel_size != 1:
+        if out_channels >= 128:
+            percent_on = 0.3
+
+    if percent_on >= 1.0:
+        return None
+    else:
+        return dict(
+            percent_on=percent_on,
+            boost_strength=1.0,
+            boost_strength_factor=0.9,
+            local=True,
+            k_inference_factor=1.0,
+        )
+
+
+def my_auto_sparse_linear_params(input_size, output_size):
+    """
+    Custom weight params.
+    :return: a dict to pass to `SparseWeights`
+    """
+    return dict(
+        weight_sparsity=0.25,
+    )
+
+
+# This configuration gets between 80.5 to 82% after 60 epochs
+SPARSE100_R1 = copy.deepcopy(DEFAULT)
+SPARSE100_R1.update(dict(
+    # No weight decay from batch norm modules
+    batch_norm_weight_decay=False,
+    init_batch_norm=True,
+
+    epochs=60,
+    checkpoint_freq=1,
+    keep_checkpoints_num=2,
+    checkpoint_score_attr="training_iteration",
+    checkpoint_at_end=True,
+    seed=tune.sample_from(lambda spec: np.random.randint(2, 10000)),
+
+    num_classes=100,
+
+    model_args=dict(config=dict(
+        num_classes=100,
+        defaults_sparse=True,
+        activation_params_func=my_auto_sparse_activation_params,
+        conv_params_func=my_auto_sparse_conv_params,
+        linear_params_func=my_auto_sparse_linear_params
+    )),
+
+    # Use a higher learning rate and no momentum for sparse superconvergence
+    lr_scheduler_class=torch.optim.lr_scheduler.OneCycleLR,
+    lr_scheduler_args=dict(
+        max_lr=6.0,
+        div_factor=6,  # initial_lr = 1.0
+        final_div_factor=4000,  # min_lr = 0.00025
+        pct_start=5.0 / 60.0,
+        epochs=60,
+        anneal_strategy="linear",
+        max_momentum=0.01,
```

Comment:
`max_momentum` is not used when `cycle_momentum` is false

Reply:
ok. I feel better leaving it in more explicitly though 😄